README.md
# Quack

[![Hex pm](http://img.shields.io/hexpm/v/quack.svg?style=flat)](https://hex.pm/packages/quack)
[![GitHub license](https://img.shields.io/github/license/azohra/Quack.svg)](https://github.com/azohra/Quack/blob/master/LICENSE.md)

A simple, yet beautiful, Elixir Logger backend for Slack

## Installation

The package can be installed by adding `quack` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:quack, "~> 0.1.1"}
  ]
end
```

Then add Quack to your list of extra applications:

```elixir
  def application do
    [
      mod: {MyApplication, []},
      extra_applications: [:logger, :quack]
    ]
  end
```

## Configuration

Configure the Slack incoming webhook:

```elixir
  config :quack, webhook_url: "https://hooks.slack.com/services/...."
```

Configure your Logger application to use Quack as a backend:

```elixir
  config :logger, backends: [:console, Quack.Logger]
```

Configure the minimum-level of logs you want exported to Slack:

```elixir
  # Options are [:debug, :info, :warn, :error]
  config :quack, level: :debug
```

Configure which metadata you want to report on:

```elixir
  # You can specify a list of containing these fields: 
  #   [
  #     :application,
  #     :module,
  #     :function,
  #     :file,
  #     :line,
  #     :pid,
  #     :crash_reason,
  #     :initial_call,
  #     :registered_name
  #   ]

  config :quack, meta: [:file, :function, :line]

  # Or you can specify absolutes such as :all, or :none
  config :quack, meta: :none
```

The complete configuration should look something like this:

```elixir
  config :quack,
    level: :info,
    meta: [:file, :function, :line],
    webhook_url: "https://hooks.slack.com/services/..."

  config :logger, backends: [:console, Quack.Logger]
```

## Examples


![debug message](docs/debug.png)
![info message](docs/info.png)
![warning message](docs/warning.png)
![error message](docs/error.png)

CHANGELOG.md
# Changelog

## Version: 0.1.1

* Decoupled the data-processing and data-transmission into seperate stages implemented by genservers. 
The module `Quack.Mama` handles the formatting and encoding of events, while `Quack.Duckling` is now only
respoinsible for hitting the webhook.

* The webhook url is no longer fetched during each request made by `Quack.Messenger`, instead it is fetched during
the startup of `Quack.Duckling` and is maintained as the state of the genserver.

## Version: 0.1.0

Initial release of Quack

test/quack_test.exs
defmodule QuackTest do
  @moduledoc false

  use ExUnit.Case
  doctest Quack
end


test/formatter_test.exs
defmodule FormatterTest do
  @moduledoc false

  use ExUnit.Case
  alias Quack.Formatter

  doctest Quack.Formatter

  @message %{
    attachments: [
      %{
        author_link: "https://github.com/azohra/quack",
        author_name: "Quack - Elixir logging for Slack",
        color: "#00B4FF",
        fields: [
          %{
            short: false,
            title: "Priority",
            value: "info"
          },
          %{
            short: false,
            title: "Timestamp",
            value: "2018-11-05 15:04:46.613"
          },
          %{
            title: "Message",
            value: "```sample_message```"
          }
        ]
      }
    ],
    text: "*Incoming Log*"
  }

  @message_with_meta %{
    attachments: [
      %{
        author_link: "https://github.com/azohra/quack",
        author_name: "Quack - Elixir logging for Slack",
        color: "#00B4FF",
        fields: [
          %{
            short: false,
            title: "Priority",
            value: "info"
          },
          %{
            short: false,
            title: "Timestamp",
            value: "2018-11-05 15:04:46.613"
          },
          %{
            title: "Metadata",
            value: "line: 17"
          },
          %{
            title: "Message",
            value: "```sample_message```"
          }
        ]
      }
    ],
    text: "*Incoming Log*"
  }

  test "properly format as bold" do
    assert Formatter.to_bold("test") == "*test*"
  end

  test "properly format as italics" do
    assert Formatter.to_italics("test") == "_test_"
  end

  test "properly format as code" do
    assert Formatter.to_code("test") == "`test`"
  end

  test "properly format as preformatted text" do
    assert Formatter.to_preformatted("test") == "```test```"
  end

  test "properly format as a quote" do
    assert Formatter.to_quote("test") == ">test"
  end

  test "properly parse a time signature" do
    ts = {{2018, 11, 5}, {15, 4, 46, 613}}

    assert Formatter.parse_ts(ts) == "2018-11-05 15:04:46.613"
  end

  test "message composition" do
    assert Formatter.create_message(
             {:info, "sample_message", {{2018, 11, 5}, {15, 4, 46, 613}}, ""}
           ) == @message

    assert Formatter.create_message(
             {:info, "sample_message", {{2018, 11, 5}, {15, 4, 46, 613}}, "line: 17"}
           ) == @message_with_meta
  end
end


mix.exs
defmodule Quack.MixProject do
  use Mix.Project

  def project do
    [
      app: :quack,
      version: "0.1.1",
      elixir: "~> 1.7",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      test_coverage: [tool: ExCoveralls],
      description: description(),
      package: package(),
      name: "Quack",
      source_url: "https://github.com/azohra/Quack"
    ]
  end

  def application do
    [
      mod: {Quack, []},
      extra_applications: [:logger]
    ]
  end

  defp deps do
    [
      {:tesla, "~> 1.2.0"},
      {:credo, "~> 0.10.2", only: [:dev, :test], runtime: false},
      {:excoveralls, "~> 0.10.2", only: [:test, :dev]},
      {:poison, ">= 1.0.0"},
      {:ex_doc, ">= 0.0.0", only: :dev}
    ]
  end

  defp description() do
    "A simple, yet beautiful, Elixir Logger backend for Slack."
  end

  defp package() do
    [
      files: ~w(lib .formatter.exs mix.exs README.md LICENSE.md CHANGELOG.md),
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/azohra/Quack"}
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :quack, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:quack, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env()}.exs"


lib/quack/logger.ex
defmodule Quack.Logger do
  @moduledoc false

  alias Quack.Mama

  @behaviour :gen_event

  @doc false
  @impl :gen_event
  def init(__MODULE__) do
    {:ok, configure()}
  end

  @doc false
  @impl :gen_event
  def init({__MODULE__, level}) do
    {:ok, configure(level)}
  end

  @doc false
  @impl :gen_event
  def handle_call(_request, state) do
    {:ok, state, []}
  end

  @doc false
  @impl :gen_event
  def handle_info(_msg, state) do
    {:ok, state}
  end

  @doc false
  @impl :gen_event
  def handle_event(:flush, state) do
    {:ok, state}
  end

  @doc false
  @impl :gen_event
  def handle_event({lvl, _gl, {_, msg, ts, meta}}, state) do
    %{level: base_level, metadata: specified} = state

    params = {lvl, msg, ts, take_params(meta, specified)}

    if valid_priority(lvl, base_level), do: Mama.process(params)

    {:ok, state}
  end

  @doc false
  @impl :gen_event
  def terminate(_reason, _state) do
    :ok
  end

  # ----------HELPER FUNCTIONS----------#

  # Function to determine if the incoming event is of valid priority
  defp valid_priority(_incoming_level, nil), do: true

  defp valid_priority(incoming_level, base_level) do
    Logger.compare_levels(incoming_level, base_level) != :lt
  end

  # Function to configure the initial application state when the user specifies
  # the desired level in init/2
  defp configure(level) do
    meta = Application.get_env(:quack, :meta)

    %{
      level: level,
      metadata: meta
    }
  end

  # Function to configure the initial application state when init/1 is called
  defp configure do
    meta = Application.get_env(:quack, :meta)
    level = Application.get_env(:quack, :level)

    %{
      level: level,
      metadata: meta
    }
  end

  # Function that takes and stringifies the given params from a keyword list
  defp take_params(_data, :none), do: ""

  defp take_params(data, :all), do: format_keyword_list(data)
  defp take_params(data, nil), do: format_keyword_list(data)

  defp take_params(data, fields) do
    data
    |> Keyword.take(fields)
    |> format_keyword_list
  end

  # Helper function that stringifies each {key, val} in a keyword list
  defp format_keyword_list(list) do
    list
    |> Enum.reduce("", fn {k, v}, acc -> "#{k}: #{inspect(v)}\n" <> acc end)
  end
end


lib/quack/formatter.ex
defmodule Quack.Formatter do
  @moduledoc """
  Module responsible for formatting and composing messages to be sent to slack
  """

  import Logger.Formatter, only: [format_date: 1, format_time: 1]

  @doc """
  Function to compose a new message based on event data
  """
  def create_message({level, message, timestamp, ""}) do
    %{
      text: "*Incoming Log*",
      attachments: [
        %{
          author_name: "Quack - Elixir logging for Slack",
          author_link: "https://github.com/azohra/quack",
          color: get_colour(level),
          fields: [
            %{
              title: "Priority",
              value: Atom.to_string(level),
              short: false
            },
            %{
              title: "Timestamp",
              value: parse_ts(timestamp),
              short: false
            },
            %{
              title: "Message",
              value: to_preformatted(message)
            }
          ]
        }
      ]
    }
  end

  def create_message({level, message, timestamp, metadata}) do
    %{
      text: "*Incoming Log*",
      attachments: [
        %{
          author_name: "Quack - Elixir logging for Slack",
          author_link: "https://github.com/azohra/quack",
          color: get_colour(level),
          fields: [
            %{
              title: "Priority",
              value: Atom.to_string(level),
              short: false
            },
            %{
              title: "Timestamp",
              value: parse_ts(timestamp),
              short: false
            },
            %{
              title: "Metadata",
              value: metadata
            },
            %{
              title: "Message",
              value: to_preformatted(message)
            }
          ]
        }
      ]
    }
  end

  @doc """
  Function to format text as code

  iex> Quack.Formatter.to_code("example")
  "`example`"
  """
  def to_code(text), do: ~s(`#{text}`)

  @doc """
  Function to format text as preformatted

  iex> Quack.Formatter.to_preformatted("example")
  "```example```"
  """
  def to_preformatted(text), do: ~s(```#{text}```)

  @doc """
  Function to format text as bold

  iex> Quack.Formatter.to_bold("example")
  "*example*"
  """
  def to_bold(text), do: ~s(*#{text}*)

  @doc """
  Function to format text as italics

  iex> Quack.Formatter.to_italics("example")
  "_example_"
  """
  def to_italics(text), do: ~s(_#{text}_)

  @doc """
  Function to format text as a quote

  iex> Quack.Formatter.to_quote("example")
  ">example"
  """
  def to_quote(text), do: ~s(>#{text})

  @doc """
  Function to format a timestamp as a string

  iex>Quack.Formatter.parse_ts({{2018, 11, 5}, {15, 4, 46, 613}})
  "2018-11-05 15:04:46.613"
  """
  def parse_ts({date, time}) do
    d =
      date
      |> format_date()
      |> to_string()

    t =
      time
      |> format_time
      |> to_string

    d <> " " <> t
  end

  # Function to return a colour based on logger level
  defp get_colour(:debug), do: "#9215E8"
  defp get_colour(:info), do: "#00B4FF"
  defp get_colour(:warn), do: "#E8BD08"
  defp get_colour(:error), do: "#FF3B0A"
end
